models/models_methods.go

```diff
@@ -111,16 +111,36 @@ func (config *BitriseDataModel) getPipelineIDs() []string {
 
 func (bundle *StepBundleModel) Normalize() error {
 	for idx, stepListItem := range bundle.Steps {
-		stepID, step, err := stepListItem.GetStepIDAndStep()
+		key, t, err := stepListItem.GetKeyAndType()
 		if err != nil {
 			return err
 		}
 
-		if err := step.Normalize(); err != nil {
-			return err
+		if t == StepListItemTypeStep {
+			step, err := stepListItem.GetStep()
+			if err != nil {
+				return err
+			}
+
+			if err := step.Normalize(); err != nil {
+				return err
+			}
+
+			stepListItem[key] = *step
+			bundle.Steps[idx] = stepListItem
+		} else if t == StepListItemTypeBundle {
+			b, err := stepListItem.GetBundle()
+			if err != nil {
+				return err
+			}
+
+			if err := b.Normalize(); err != nil {
```

Comment:
I think we should have a recursion limit here. It's too easy to accidentally create infinite recursion with bundle references, and if we don't detect it early (here), then in the worst case scenario, someone DoSes the Monolith with a build that tries to validate an infinitely recursive bitrise.yml 😆

Reply:
step bundle circular reference is validated here: https://github.com/bitrise-io/bitrise/pull/1050/commits/0a88e7016171b87735c98e50819bbf4dbe7ea8b7